Add intro page on APIs (#483)
giantswarm/docs: src/content/basics/api/index.md

---
title: API Access to Giant Swarm Resources
description: An overview of the APIs that provide you with programmatic access to
  resources like your tenant cluster in a Giant Swarm installation. Namely the Control
  Plane Kubernetes API and the legacy Giant Swarm Rest API.
date: 2020-05-04
weight: 75
type: page
categories: ["basics"]
last-review-date:
---

# API Access to Giant Swarm Resources

For integrating Giant Swarm with your automation, we provide programmatic access via two different APIs:

> - [The Control Plane Kubernetes API](#cp-k8s-api)

Reviewer: I'd turn this around.

The Control Plane API is in preview
The Giant Swarm Rest API is stable

Author: Ack

- [The Giant Swarm Rest API](#rest-api) (deprecated)

In the following sections we explain the differences and specific benefits and provide some guidance on how to get started.

## The Control Plane Kubernetes API (CP-K8s-API) {#cp-k8s-api}

## What it is

> At Giant Swarm, when we say "control plane", we talk about the Kubernetes Cluster that runs all the operational and monitoring workloads which are needed to create and manage the "tenant clusters", which are the clusters you create to run your actual workloads. This is not to be confused with how the Kubernetes project uses the same term, simply meaning the master nodes of a Kubernetes cluster.

Reviewer 2: ```suggestion
At Giant Swarm, when we say "control plane", we mean the Kubernetes Cluster that runs all the operational and monitoring workloads which are needed to create and manage the "tenant clusters". These are the clusters you create to run your actual workloads. This is not to be confused with how the Kubernetes project uses the same term. In that case, it simply refers to the master nodes of a Kubernetes cluster.
```

> The control plane is a Kubernetes cluster. Your tenant clusters and other associated resources are represented in that cluster as custom resources. To access these, you can use the Kubernetes API of the cluster that forms the control plane, or in short, the Control Plane Kubernetes API. For the sake of brevity, we'll abbreviate this as _CP-K8s-API_.

Reviewer: But in the long run this API will be called the Giant Swarm API, or not?

_CP-K8s-API_ sounds very clunky. Any other ideas? Why not just _CP-API_?

Reviewer 3: Should the final name not be `Giant Swarm Control Plane API`?
I personally think we should restrain from using abbreviation when we talk to customers (public doc and support channels)
Is there any reason to not call it Control Plane API in our public doc?

Author: > Is there any reason to not call it Control Plane API in our public doc?

Maybe. Not sure it is useful to have the term "Kubernetes" in there.

Author: Settling with "Control Plane Kubernetes API" for now.

### How to gain access

> To access the CP-K8s-API, Giant Swarm configures access to the API via OIDC. Please contact Solution Engineer (SE) to sort out the details.

Reviewer 3: ```suggestion
Access to the Giant Swarm Control Plane API is secured using OIDC. Please contact your Solution Engineer (SE) to sort out the details.
```

> Currently we normally provide read-only access by default. As we are currently working on a fine-grained way to control permissions to resources, having write access right now means fully unrestricted access and should only be granted to select individuals.

Reviewer: ```suggestion
Currently we provide read-only access by default. As we are currently working on a fine-grained way to control permissions to resources, having write access right now means fully unrestricted access and should only be granted to select individuals.
```

Author: Ack

> TODO: Explain how setting up write is different from read and what else it requires.

Reviewer: Validation and defaulting is not there yet and this will be another pitfall for users.

Author: Ack

### How to use

> We recommend that you become familiar with the API via `kubectl`, navigating your resources via the CP-K8s-API. Besides general Kubernetes know-how this will require only a bit of structural knowledge:

Reviewer 4: Can we condense this sentence a bit? Also could we recommend something lik eoctant here as a way to visually navigate and explore through the resources?

```suggestion
We recommend using `kubectl` to navigate the resources present on the CP-K8s-API. Besides general Kubernetes know-how this will require only a bit of structural knowledge:
```

#### How we organize resources in namespaces

> This is quickly explained: we create one namespace for each tenant cluster, where the namespace name is equal to the tenant cluster ID. All cluster specific resources reside in the namespace of that tenant cluster.

Reviewer 2: ```suggestion
We create one namespace for each tenant cluster, where the namespace name is equal to the tenant cluster ID. All cluster specific resources reside in the namespace of that tenant cluster.
```

> ### Which custom resources we use for what purpose

Reviewer 2: ```suggestion
### Which custom resources are used for what purpose?
```

Author: Ack, minus the question mark. I don't think punctuation belongs into headlines

> For this, we provide some resources that should help you:

Reviewer 2: ```suggestion
Following are some resources that should help you:
```

- The guide [Creating tenant clusters via the Control Plane Kubernetes API](/guides/creating-clusters-via-crs-on-aws/) explains step by step how you can create a cluster and node pools via the CP-K8s-API. Here you learn about all the custom resources a cluster comprises.
- The [App Platform](/basics/app-platform/) introduction outlines the several custom resources involved when managing app catalogs and apps.
- Our [Control Plane Kubernetes API Reference](https://docs.giantswarm.io/reference/cp-k8s-api/) provides detailed documentation on all the custom resources we use with the various providers and their versions and schema.

### Feedback is welcome

We are keen to learn from you about your experience with accessing the CP-K8s-API, with navigating the custom resources via the API, with our reference documentation and the user guides we provide. This helps us provide more and better material and improve to make the journey more seamless, more satisfactory for you.

So please, don't hesitate to give your feedback in your Slack channel.

> ## The Giant Swarm Rest API (deprecated) {#rest-api}

Reviewer: I'd still move this to the top for now. This should be the API that first users eg if we onboard customers should use.

Reviewer: Hm. If this comes first the description also doesn't fit anymore.

Reviewer 2: +1 for starting with this part, while making the necessary adjustments to the new content.

Author: Will do

The control plane is a Kubernetes cluster. Your tenant clusters and other associated resources are represented in that cluster as [custom resources](https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/). To access these, you can use the Kubernetes API of the cluster that forms the control plane, or in short, the Control Plane Kubernetes API.

### How to gain access

Access to the Giant Swarm Control Plane API is secured using OIDC. Please contact your Solution Engineer (SE) to sort out the details.

Currently we provide read-only access by default. As we are currently working on a fine-grained way to control permissions to resources, and validation and defaulting are not implemented to the extend we want to, having write access right now means fully unrestricted access and should only be granted to select individuals.

> TODO: Explain how setting up write is different from read and what else it requires.

Reviewer 4: There's a todo here, I'm fine removing it for now and making a issue out of it so we get back to it later.

Author: Done
